pkg/pointerdb/pointerdb.go

```diff
@@ -228,3 +231,70 @@ func (s *Server) Delete(ctx context.Context, req *pb.DeleteRequest) (resp *pb.De
 	s.logger.Debug("deleted pointer at path: " + req.GetPath())
 	return &pb.DeleteResponse{}, nil
 }
+
+func (s *Server) iterate(ctx context.Context, req *pb.IterateRequest, queue dr.Queue) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	s.logger.Debug("entering pointerdb iterate")
+	
+	if err = s.validateAuth(req.GetAPIKey()); err != nil {
+		return err
+	}
+	
+	iter := func(it storage.Iterator) error {
+		var item storage.ListItem
+		for ; req.Limit > 0 && it.Next(&item); req.Limit-- {
+			if item.Key == nil {
+				panic("nil key")
+			}
+			pointer := &pb.Pointer{}
+			err = proto.Unmarshal(item.Value, pointer)
+			if err != nil {
+				//TODO better error handling
+				return err
+			}
+			pieces := pointer.Remote.RemotePieces
+			var nodeIDs []dht.NodeID
+			for i, p := range pieces {
+				nodeIDs = append(nodeIDs, kademlia.StringToNodeID(p.NodeId))
+			}
+			missingPieces, err := offlineNodes(ctx, nodeIDs)
```

Comment:
The list of offline nodes should be cached, a lookup per pointer is very inefficient.

Reply:
from an architectural perspective, this is precisely why the overlay cache exists. the overlay cache exists to make lookups per pointer faster. do we want a cache on top of the cache?

Reply:
Ah, right forgot that there is a cache. However, it might get contended a lot and a local map of offline nodes, might be helpful. But, yeah... leave it as is for now.

Reply:
yeah, it is a genuine question in terms of if we want a local cache on top for less contention. might be a good call, but also feels like it might be closer on the spectrum to premature optimization